src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests.vb

```diff
@@ -3429,61 +3429,174 @@ class C
+                    Await Task.Delay(250)
 
-                Await state.WaitForAsynchronousOperationsAsync()
-                Await state.AssertNoCompletionSession()
-                Assert.Contains("System", state.GetLineTextFromCaretPosition())
+                    state.AssertNoCompletionSessionWithNoBlock()
+                    Assert.Contains("Sys", state.GetLineTextFromCaretPosition())
+                    Assert.DoesNotContain("System", state.GetLineTextFromCaretPosition())
+                    tcs.SetResult(True)
+
+                    Await state.WaitForAsynchronousOperationsAsync()
+                    Await state.AssertNoCompletionSession()
+                    Assert.Contains("System", state.GetLineTextFromCaretPosition())
+                Else
+                    Dim task1 As Task = Nothing
+                    Dim task2 As Task = Nothing
+
+                    Dim providerCalledHandler = Sub()
+                                                    task1 = Task.Run(Sub()
+                                                                         ' 2. Hang here as well.
+                                                                         Dim completionItem = state.GetSelectedItemOpt()
+                                                                     End Sub)
+                                                    task2 = Task.Run(Sub()
+                                                                         Thread.Sleep(250)
+                                                                         Try
+                                                                             ' 3. Check that the other task is running/hanging.
+                                                                             Assert.Equal(TaskStatus.Running, task1.Status)
+                                                                             Assert.Contains("Sys", state.GetLineTextFromCaretPosition())
+                                                                             Assert.DoesNotContain("System", state.GetLineTextFromCaretPosition())
+                                                                         Finally
+                                                                             ' 4. Unblock the first task and the main thread.
+                                                                             tcs.SetResult(True)
+                                                                         End Try
+                                                                     End Sub)
+                                                End Sub
+
+                    Try
+                        AddHandler provider.ProviderCalled, providerCalledHandler
+
+                        ' In the new completion, when pressed <ctrl>-<space>, we have to wait for the aggregate operation to complete.
+                        ' 1. Hang here.
+                        state.SendCommitUniqueCompletionListItem()
```

Comment:
Yes, it guarantees because SendCommitUniqueCompletionListItem is a synchronous operation. Added a comment.

---
In reply to: [247000131](https://github.com/dotnet/roslyn/pull/31534#discussion_r247000131) [](ancestors = 247000131)